initial draft (#7)
SteveL-MSFT/PowerShell-RFC: 2-Draft-Accepted/RFCxxxx-DSC-for-PowerShell7.md

---
RFC: RFCxxxx
Author: Steve Lee
Status: Draft
SupercededBy: N/A
Version: 1.0
Area: Desired State Configuration
Comments Due: Nov 22, 2020
Plan to implement: Yes
---

# DSC for PowerShell 7

[Desired State Configuration](https://docs.microsoft.com/powershell/scripting/dsc/overview/overview) (DSC) is a platform in PowerShell

> enabling configuration as code.

Reviewer: I feel like this is a missed opportunity to be a bit more exhaustive on the components and the roles involved in the DSC ecosystem (DSL, DSC Resources, Configuration Management tool for Windows with the LCM, Pull Server, a way to imperatively consume the DSC Resources with Invoke-DscResources).
By naming and differentiating those, it's easier to talk about what will and what won't be invested on going forward.
The lack of differentiation under that broad "DSC" has been detrimental in the past as it was misunderstood.

Author: I'll add some text to clarify what is part of this RFC and what is not, but didn't want to duplicate the content in the link provided and also expect anyone reading this RFC to be familiar with DSC.

Currently, DSC (for Windows PowerShell) is tightly coupled with WMI on Windows and OMI on Linux.
WMI is used to host the Local Configuration Manager (LCM) and WMI APIs are used to parse MOF which is used for both
configuration files as well as schema files.

In addition, it is not possible to write DSC resources that are cross platform compatible nor easily author DSC resources on Linux as
they need to be in C++ or Python.
Finally, third party solutions trying to leverage DSC resources need to still go through the LCM making it complicated to integrate with
their agents or ecosystems.

> Unlike DSC for Windows PowerShell, DSC for PowerShell 7 is only a platform for building or integrating with configuration solutions.

Reviewer: [...]DSC for PowerShell 7 only intend to provide a platform for building or integrating with configuration solutions.

## Motivation

    As a developer of configuration solutions,
    I can easily leverage DSC resources,

>     so that my solution can manage Windows.

Reviewer: "so that my solution can manage Windows." too restrictive IMO.
The real benefit as always to tap into the PowerShell ecosystem, community and consumers.
"so that my solution can leverage the PowerShell ecosystem and its users and contributors".

Instead, we want to make it easier to integrate with existing configuration agents like [Ansible](https://docs.ansible.com/ansible/latest/user_guide/windows_dsc.html),
[Chef](https://docs.chef.io/resources/dsc_resource/), [Puppet](https://forge.puppet.com/puppetlabs/dsc),
[Azure Policy Guest Configuration](https://docs.microsoft.com/azure/governance/policy/concepts/guest-configuration), and other popular configuration solutions.

Custom agents can simply use the existing PowerShell API (or call out to `pwsh`) to directly invoke DSC resources.

You can even simply write a PowerShell script as your agent that runs off TaskScheduler or cron, for example.

Along with the lack of a default agent, there is no pre-defined push or pull model for deploying DSC.

### Cross platform first

DSC has supported both Windows and Linux, however, the [Linux support](https://docs.microsoft.com/powershell/scripting/dsc/getting-started/lnxGettingStarted?view=powershell-7)
was dependent on [OMI](https://github.com/Microsoft/omi) and did not support authoring Linux DSC resources in PowerShell script making development
inconsistent and much more difficult.

> DSC for PowerShell 7 will be cross platform having consistent capabilities and user experience on all platforms.

Reviewer: what's the plan with python, c++ and other resource "types"?

Author: Not supported at this time

Author: I'll add some text about this to make it clear

### Moving away from WMI/OMI/MOF

DSC on Windows was originally designed relying on [WMI](https://docs.microsoft.com/windows/win32/wmisdk/wmi-start-page) as a host on Windows,

> OMI as a host on Linux, and [MOF](https://docs.microsoft.com/windows/win32/wmisdk/managed-object-format--mof-) for configuratio and schema.

Reviewer: typo: "for configuratio**n** and schema".
consider mentioning CIM is involved to convert PS Data Structures back in forth, even when invoking from PowerShell.

Author: will add some language about that

MOF as a format is not something most developers are familiar with and not much tooling support.
However, developers of DSC resources written in PowerShell script still needed to learn MOF and also keep their MOF schema
in sync with their script implementation.

Today, when you execute a [DSC configuration script](https://docs.microsoft.com/powershell/scripting/dsc/configurations/configurations?view=powershell-7),
the output is a MOF configuration file.
In general, users do not need to read this MOF file as it's intended to be used by machines.
However, MOF parsing creates a dependency on WMI/OMI.

> Instead, we will move towards a JSON configuration file format.

Reviewer: The serialized JSON file format is not the most critical thing IMO.
Having access to the in-memory graph of resource, and if possible on the fly during compilation, can enable some nice scenarios around config data (i.e. JIT credential fetching from vault).

The serialized version of this is great for a common data serialization structure, but we've already "lost" some info (metadata about config data).

Author: Additional tooling to understand the dependency graph will have to come later.  We have a partner already that wants the JSON format as they are building their own LCM.  At this point, I think the scope is to define the JSON format and leave it to developers to handle the rest.

Reviewer 2: Whose responsibility will it be to make sure that PSCredential objects don't end-up in plain text in the compiled JSON file if we are to bring our own agent? How do we foresee credential encryption work here?

Since there is no default agent to understand the JSON configuration file, it will be up to users or 3rd party agents to understand the configuration file.

> 

Reviewer: because of comment above, I'd remove this sentence and JSON config example, that should probably be a different RFC in itself.

<TODO> example JSON configuratoin </TODO>

### No more MOF schema and script resources

Although DSC resources can be written as WMI/OMI providers, this requires writing C/C++ code and compiling for specific Linux distros and processor architectures.
Writing PowerShell script DSC resources currently still requires a matching schema file making it more complex.
Instead, the current plan is to [only support PowerShell class based DSC resources](https://github.com/PowerShell/PowerShell/issues/13731) in DSC for PowerShell 7.
The reasoning and discussion is in the issue linked above.

### Open Source

The current [PSDesiredStateConfiguration](https://docs.microsoft.com/powershell/module/psdesiredstateconfiguration/?view=powershell-7) module is not Open Source.

> The intent is to clean up the code and prepare the GitHub repository to be made public as an Open Source project.

Reviewer: You say clean up the code, but you don't mention touching the DSL... Intentional omission?

### PowerShell engine changes

To enable DSC for PowerShell 7 to evolve independently to PowerShell 7 itself, the code in the engine that handles the DSC configuration keyword will
be seperated as a plugin and made to be part of the PSDesiredStateConfiguration module.

## Alternate Proposals and Considerations

### Supporting existing script based resources

Existing script based resources (not class based) require a MOF schema file for PowerShell to understand the types and nested types.
We had prototyped using a JSON based schema instead of the existing MOF schema file, but it still required the DSC resource author to
create a schema file and ensure it aligns with the script implementation.

> It is simpler to create a PoewrShell class based resource which itself is the schema removing the need to understand and maintain a

Reviewer: typo on PowerShell

separate schema file.
In addition, having a single model (class based) for authoring DSC resources will help the community in the long term by having a single

> consistent way to write DSC resources making it easier for new authors to find documentation and examples.

Reviewer: consider linking to this article: https://dsccommunity.org/blog/class-based-dsc-resources/

Author: will do

WMI is used to host the Local Configuration Manager (LCM) and WMI APIs are used to parse MOF which is used for both
configuration files as well as schema files.

In addition, it is not possible to write DSC resources that are cross platform compatible nor easily author DSC resources on Linux as
they need to be in C++ or Python.
Finally, third party solutions trying to leverage DSC resources need to still go through the LCM making it complicated to integrate with
their agents or ecosystems.

Unlike DSC for Windows PowerShell, DSC for PowerShell 7 is only intended to provide a platform for building or integrating with configuration solutions.
This means for PowerShell 7, the scope of DSC support is limited to `configuration` keyword and compliation, authoring and using
DSC resources via the PSDesiredStateConfiguration module.

## Motivation

    As a developer of configuration solutions,
    I can easily leverage DSC resources,
    so that my solution can leverage the PowerShell ecosystem and its users and contributors.

    As an ITPro,
    I can directly invoke DSC resources,
    so that I can build ad hoc cross platform configuration.

> 

Reviewer 2: What would be my motivation as a DSC Module developer? Will it also make it easier for me to write DSC modules that are cross-platform and work on Windows and Linux? (I know the answer is yes, but might be good to also indicate it)

Author: This section was updated by Michael, please re-review

inconsistent and much more difficult.

DSC for PowerShell 7 will be cross platform having consistent capabilities and user experience on all platforms.

### Moving away from WMI/OMI/MOF

DSC on Windows was originally designed relying on [WMI](https://docs.microsoft.com/windows/win32/wmisdk/wmi-start-page) as a host on Windows,
OMI as a host on Linux, and [MOF](https://docs.microsoft.com/windows/win32/wmisdk/managed-object-format--mof-) for configuration and schema.

MOF as a format is not something most developers are familiar with and not much tooling support.
However, developers of DSC resources written in PowerShell script still needed to learn MOF and also keep their MOF schema
in sync with their script implementation.

Today, when you execute a [DSC configuration script](https://docs.microsoft.com/powershell/scripting/dsc/configurations/configurations?view=powershell-7),
the output is a MOF configuration file.
In general, users do not need to read this MOF file as it's intended to be used by machines.
However, MOF parsing creates a dependency on WMI/OMI.
Instead, we will move towards a JSON configuration file format.

> Since there is no default agent to understand the JSON configuration file, it will be up to users or 3rd party agents to understand the configuration file.

Reviewer 2: May be a great potential to Open-Source whatever parser we come up with for the Azure Guest Config agent to use as a poster child for the community in order to encourage contributions and help people wanting to build their own get a something to start with. (Food for thoughts)

Author: I'll add some content to the `Considerations` section to cover this as something we could do in the future

---
RFC: RFCxxxx
Author: Steve Lee
Status: Draft
SupercededBy: N/A
Version: 1.0
Area: Desired State Configuration
Comments Due: Nov 22, 2020
Plan to implement: Yes
---

# DSC for PowerShell 7

[Desired State Configuration](https://docs.microsoft.com/powershell/scripting/dsc/overview/overview)
(DSC) is a platform in PowerShell enabling configuration as code.

## Background

DSC contains many components including:

 - The `configuration` keyword that is part of the PowerShell engine
 - Compilation to a configuration MOF file
 - DSC resources (those that ship in Windows and those on PowerShellGallery.com)

>  - Local Configuration Manager in Windows and maintained as open-source for Linux

Reviewer 3: ```suggestion
 - Local Configuration Manager (LCM) in Windows and maintained as open-source for Linux,
   a service that maintains and/or reports on configuration drift
```

>  - Pull Server support

Reviewer 3: ```suggestion
 - the DSC Pull Server, which is used by LCM's configured in pull mode to acquire new
   resources and configurations
```

>  - PSDesiredStateConfiguration module.

Reviewer 3: ```suggestion
 - the `PSDesiredStateConfiguration` module that includes cmdlets like `Get-DscResource`
```

DSC ships as a component of Windows and as an open source project for Linux.

> It is also integrated in to cloud-based server management services in multiple

Reviewer 3: ```suggestion
It is also integrated into cloud-based server management services in multiple
```

> public clouds including Microsoft Azure and Amazon AWS.

Reviewer 3: ```suggestion
public clouds including Microsoft Azure and Amazon Web Services (AWS).
```

The versions of DSC that shipped in Windows PowerShell are tightly coupled with
[Windows Management Instrumentation (WMI)](https://docs.microsoft.com/windows/win32/wmisdk/wmi-start-page)
on Windows and
[Open Management Interface](https://www.dmtf.org/content/open-management-interface-omi-%E2%80%93-open-source-implementation-dmtf-cimwbem-standards)
(OMI) on Linux.
The *MI libraries are used to host the Local Configuration Manager (LCM) and APIs
are used to parse information store in Managed Object Framework (MOF) file format,
which is used for both configuration files and schema files.

### Challenges with the current versions of DSC

> Very few community authored DSC resources exist for Linux because C++ is the only

Reviewer 3: Isn't Python also supported? Or did we never officially support that?

Author: Michael asked the same question.  I thought it was supported, but not aware of any resources built on it.

Reviewer 4: From a convo with Norberto a couple years ago, I thought that _at a minimum_ the nxFileLine resource was built with Python.

Reviewer 4: Ahhh, searched my email.  It was a Python USE.  The resource was still built in CPP.

supported language. Other than using the Script/nxScript resources combined with

> a composite resource, it is near-impossible to create a DSC solution that is cross-platform.

Reviewer 3: ```suggestion
a composite resource, it is near-impossible today to create a DSC solution that is cross-platform.
```

> Third party solutions trying to leverage DSC resources need to call LCM

Reviewer 3: ```suggestion
Today, third party solutions trying to leverage DSC resources need to call the LCM
```

to run Get/Set/Test, making it complicated to integrate with partner ecosystems.

## Scope of DSC implementation for PowerShell 7

Unlike DSC for Windows PowerShell, DSC for PowerShell 7 is only intended to provide
a platform for building or integrating with configuration solutions. For PowerShell 7,

> the scope of DSC support is limited to `configuration` keyword, authoring,

Reviewer 3: ```suggestion
the scope of DSC support is limited to the `configuration` keyword, authoring,
```

compilation, and using DSC resources via the PSDesiredStateConfiguration module.

> 

Reviewer 3: Can we add what's not included? I.e. LCM and pull server

Author: Perhaps we can make it explicit

### Motivation

> DCS in PowerShell 7 should enable the following user scenarios.

Reviewer 3: ```suggestion
DSC in PowerShell 7 should enable the following user scenarios:
```

    As a developer of configuration management solutions,
    I can easily integrate DSC resources
    so that my solution can leverage the PowerShell ecosystem and its users and contributors.

    As an IT Pro,
    I can directly invoke DSC resources
    so that I can build ad hoc cross platform configurations.

### Cross platform first

DSC has supported both Windows and Linux, however, the
[Linux support](https://docs.microsoft.com/powershell/scripting/dsc/getting-started/lnxGettingStarted?view=powershell-7)
was dependent on a Microsoft-maintained implementation of
[OMI](https://github.com/Microsoft/omi)
and did not encourage authoring Linux DSC resources in PowerShell,
making development inconsistent and difficult for non-expert users.

> DSC for PowerShell 7 will prioritize being cross platform, having consistent
> capabilities and user experience on all supported platforms.

Reviewer 3: ```suggestion
DSC for PowerShell 7 will prioritize being cross platform and having a consistent
user experience and capabilities on all supported platforms.
```

      ]
    }
  ],
  "FileDirectoryConfiguration": [
    {
      "ResourceID": "[File]DirectoryCopy",
      "Ensure": "Present",
      "Recurse": true,
      "ModuleVersion": "1.0",
      "DestinationPath": "C:\\\\Users\\\\Public\\\\Documents\\\\DSCDemo\\\\DemoDestination",
      "ConfigurationName": "FileResourceDemo",
      "SourceInfo": "C:\\\\Users\\\\slee\\\\test\\\\file.dsc.ps1::6::9::File",
      "ModuleName": "PSDesiredStateConfiguration",
      "Type": "Directory",
      "SourcePath": "\\\\\\\\PullServer\\\\DemoSource"
    }
  ]
}
```

Important differences:

- CIM class name prefixes are removed as they are not needed
- Instances are represented as an array of the type name
- The comment at the start is dropped as that information is available within `ConfigurationDocument`

### No longer require MOF schema by switching to improved class-based resources

Although DSC resources can be written as WMI/OMI providers, this requires writing
C/C++ code and compiling for specific Linux distros and processor architectures.
Writing PowerShell script DSC resources currently still requires a matching schema
file making it more complex. Instead, the current plan is to

> [only support PowerShell class based DSC resources](https://github.com/PowerShell/PowerShell/issues/13731)

Reviewer 3: ```suggestion
[only support PowerShell class-based DSC resources](https://github.com/PowerShell/PowerShell/issues/13731)
```

in DSC for PowerShell 7. Whereas previously embedded objects were CIMInstance types,

> they would now be .NET types based on PowerShell classes. Additional reasoning

Reviewer 3: ```suggestion
they will now be .NET types based on PowerShell classes. Additional reasoning
```

and discussion is in the issue linked above.

> Making the change from script based resources to class based resources could make

Reviewer 3: ```suggestion
Making the change from script-based resources to class-based resources could make
```

learning DSC a frustrating experience for new users. To resolve this, tooling will

> be needed that scaffolds as much as possible of the class requirements based

Reviewer 3: ```suggestion
be needed that scaffolds as much of the class requirements as possible based
```

on a .psm1 module file containing traditionally authored Get/Set/Test-TargetResource
functions. This would also aid in migrating the more than 1000 existing resources
for Windows.

### Open Source

The current
[PSDesiredStateConfiguration](https://docs.microsoft.com/powershell/module/psdesiredstateconfiguration/?view=powershell-7)
module is not Open Source. The intent is to clean up the code and prepare the GitHub

> repository to be made public as an Open Source project.

Reviewer 3: "open source" is usually lower-case, but it's up to you here.

Also might be good to include a non-ETA "though we don't currently have a date for this work" or whatever

### PowerShell engine changes

To enable DSC for PowerShell 7 to evolve independently to PowerShell 7 itself, the
code in the engine that handles the DSC configuration keyword will be separated
as a plugin and made to be part of the PSDesiredStateConfiguration module.

## Alternate Proposals and Considerations

### Supporting existing script based resources

> Existing script based resources (not class based) require a MOF schema file for

Reviewer 3: ```suggestion
Existing script-based resources (not class-based) require a MOF schema file for
```

---
RFC: RFCxxxx
Author: Steve Lee
Status: Draft
SupercededBy: N/A
Version: 1.0
Area: Desired State Configuration
Comments Due: Nov 22, 2020
Plan to implement: Yes
---

> # DSC for PowerShell 7

Reviewer 4: Do we need to make it clear that it's not "any 7" per se, but `7.x` (not being 7.0 or 7.1)?  "for PowerShell 7" is used throughout, and in Docs I'd be worried it might infer for an already existing version of 7 since it's not fully explicit.  If this is the normal titling for the RFCs in PS, just ignore this comment. :)

Author: The `PowerShell 7` was deliberate as part of our branding and not committing to a specific version of when this work will be complete

[Desired State Configuration](https://docs.microsoft.com/powershell/scripting/dsc/overview/overview)
(DSC) is a platform in PowerShell enabling configuration as code.

## Background

DSC contains many components including:

- The `configuration` keyword that is part of the PowerShell engine
- Compilation to a configuration MOF file
- DSC resources (those that ship in Windows and those on PowerShellGallery.com)
- Local Configuration Manager (LCM) in Windows and maintained as open-source for Linux,
  a service that maintains and/or reports on configuration drift
- the DSC Pull Server, which is used by LCM's configured in pull mode to acquire new
  resources and configurations
- the `PSDesiredStateConfiguration` module that includes cmdlets like `Get-DscResource`

> DSC ships as a component of Windows and as an open source project for Linux.

Reviewer 4: ```suggestion
DSC ships as a component of Windows 10 and as an open source project for Linux.
```

It is also integrated into cloud-based server management services in multiple
public clouds including Microsoft Azure and Amazon Web Services (AWS).

> The versions of DSC that shipped in Windows PowerShell are tightly coupled with

Reviewer 4: ```suggestion
The versions of DSC that ship in Windows PowerShell are tightly coupled with
```

[Windows Management Instrumentation (WMI)](https://docs.microsoft.com/windows/win32/wmisdk/wmi-start-page)
on Windows and
[Open Management Interface](https://www.dmtf.org/content/open-management-interface-omi-%E2%80%93-open-source-implementation-dmtf-cimwbem-standards)
(OMI) on Linux.
The *MI libraries are used to host the Local Configuration Manager (LCM) and APIs
are used to parse information store in Managed Object Framework (MOF) file format,
which is used for both configuration files and schema files.

### Challenges with the current versions of DSC

Very few community authored DSC resources exist for Linux because C++ is the only
supported language. Other than using the Script/nxScript resources combined with
a composite resource, it is near-impossible today to create a DSC solution that is cross-platform.

> Today, third party solutions trying to leverage DSC resources need to call the LCM

Reviewer 4: ```suggestion
Today, third party solutions leveraging DSC resources need to call the LCM
```

to run Get/Set/Test, making it complicated to integrate with partner ecosystems.

## Scope of DSC implementation for PowerShell 7

Unlike DSC for Windows PowerShell, DSC for PowerShell 7 is only intended to provide
a platform for building or integrating with configuration solutions. For PowerShell 7,
the scope of DSC support is limited to the `configuration` keyword, authoring,
compilation, and using DSC resources via the PSDesiredStateConfiguration module.

> This means that a default agent (LCM) is not included and existing pull servers is not supported.

Reviewer 4: ```suggestion
This means that a default agent (LCM) is not included and existing pull servers are not supported.
```

### Motivation

DSC in PowerShell 7 should enable the following user scenarios:

    As a developer of configuration management solutions,
    I can easily integrate DSC resources
    so that my solution can leverage the PowerShell ecosystem and its users and contributors.

    As an IT Pro,
    I can directly invoke DSC resources
    so that I can build ad hoc cross platform configurations.

### Cross platform first

DSC has supported both Windows and Linux, however, the

> [Linux support](https://docs.microsoft.com/powershell/scripting/dsc/getting-started/lnxGettingStarted?view=powershell-7)

Reviewer 4: ```suggestion
[Linux support](https://docs.microsoft.com/powershell/scripting/dsc/getting-started/lnxGettingStarted)
```

was dependent on a Microsoft-maintained implementation of
[OMI](https://github.com/Microsoft/omi)
and did not encourage authoring Linux DSC resources in PowerShell,
making development inconsistent and difficult for non-expert users.

DSC for PowerShell 7 will prioritize being cross platform and having a consistent
user experience and capabilities on all supported platforms.

### Invoke-DscResource

About a year ago, we worked with
[Gael Colas](https://twitter.com/gaelcolas)
to author an RFC describing fundamental changes to

> [`Invoke-DscResource`](https://docs.microsoft.com/powershell/module/psdesiredstateconfiguration/invoke-dscresource?view=powershell-7)

Reviewer 4: ```suggestion
[`Invoke-DscResource`](https://docs.microsoft.com/powershell/module/psdesiredstateconfiguration/invoke-dscresource)
```

in PowerShell 7.0.
This enabled executing the Get, Set, and Test methods on DSC resources bypassing
the LCM on Windows and removing the need for OMI on Linux. This was the start
to enable cross platform DSC resources to be authored in PowerShell script (with
some limitations).

The changes to DSC shipped as an

> [experimental feature](https://docs.microsoft.com/powershell/scripting/learn/experimental-features?view=powershell-7)

Reviewer 4: ```suggestion
[experimental feature](https://docs.microsoft.com/powershell/scripting/learn/experimental-features)
```

and other popular configuration solutions.

Custom agents can use the existing PowerShell API (or call out to `pwsh`) to
directly invoke DSC resources. For ad hoc configuration management, solutions can
even use a PowerShell script run by TaskScheduler or cron.

Along with the lack of a default agent, there is no pre-defined push or pull
model for deploying DSC.

### Move away from WMI/OMI/MOF

DSC on Windows was originally designed relying on
[WMI](https://docs.microsoft.com/windows/win32/wmisdk/wmi-start-page)
as a host on Windows,
[OMI](https://www.dmtf.org/content/open-management-interface-omi-%E2%80%93-open-source-implementation-dmtf-cimwbem-standards)
as a host on Linux, and
[MOF](https://docs.microsoft.com/windows/win32/wmisdk/managed-object-format--mof-)
for configuration and schema.

MOF as a format is not something most developers are familiar with and there is no
tooling support. However, developers of DSC resources written in PowerShell script
still needed to learn MOF and also keep their MOF schema in sync with their script
implementation.

Today, when you execute a

> [DSC configuration script](https://docs.microsoft.com/powershell/scripting/dsc/configurations/configurations?view=powershell-7),

Reviewer 4: ```suggestion
[DSC configuration script](https://docs.microsoft.com/powershell/scripting/dsc/configurations/configurations),
```

### No longer require MOF schema by switching to improved class-based resources

Although DSC resources can be written as WMI/OMI providers, this requires writing
C/C++ code and compiling for specific Linux distros and processor architectures.
Writing PowerShell script DSC resources currently still requires a matching schema
file making it more complex. Instead, the current plan is to
[only support PowerShell class-based DSC resources](https://github.com/PowerShell/PowerShell/issues/13731)
in DSC for PowerShell 7. Whereas previously embedded objects were CIMInstance types,
they will now be .NET types based on PowerShell classes. Additional reasoning
and discussion is in the issue linked above.

Making the change from script-based resources to class-based resources could make
learning DSC a frustrating experience for new users. To resolve this, tooling will
be needed that scaffolds as much of the class requirements as possible based
on a .psm1 module file containing traditionally authored Get/Set/Test-TargetResource
functions. This would also aid in migrating the more than 1000 existing resources
for Windows.

### Open Source

The current

> [PSDesiredStateConfiguration](https://docs.microsoft.com/powershell/module/psdesiredstateconfiguration/?view=powershell-7)

Reviewer 4: ```suggestion
[PSDesiredStateConfiguration](https://docs.microsoft.com/powershell/module/psdesiredstateconfiguration/)
```
